### PR TITLE
[MXNET-1198] MXNet Java API

### PR DESCRIPTION
**Pull request by lanking520**
## Description ##
This PR is aimed to introduce the new Java API for MXNet. It contains the following features:
- Core package: Shape, Context, DType, DataDesc, NDArray
- Infer package: Predictor, Object Detector
- Example package: Object Detection Example, Image Classification using Predictor
Apart from those functions, the new Java API introduces a new way to use operators:
```
NDArray.dot(new dotParam(nd1, nd2).setOut(nd3))
```
Users would be able to create a Parameter object to work on operators with large optional parameters. We also expect to deliver more features to Java API such as Image classifier, Module API.

## Java contributors:
Contributors: @lanking520 @andrewfayres @piyushghai @zachgk
Comitters: @nswamy @yzhliu @gigasquid 

## Call for reviewing:
Contributors: @mdespriee  @CodingCat @Ldpe2G
Comitters: @terrytangyuan @yanqingmen 

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [x] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### CODE REVIEW

**Comment by anirudhacharya** [comment]
@mxnet-label-bot [pr-work-in-progress]

File: scala-package/examples/scripts/infer/objectdetector/run_ssd_example.sh [hunk 1]
```diff
@@ -17,9 +17,21 @@
...
+then
+    platform=osx-x86_64
+fi
 
 MXNET_ROOT=$(cd "$(dirname $0)/../../../../../"; pwd)
-CLASS_PATH=$MXNET_ROOT/scala-package/assembly/osx-x86_64-cpu/target/*:$MXNET_ROOT/scala-package/examples/target/*:$MXNET_ROOT/scala-package/examples/target/classes/lib/*:$MXNET_ROOT/scala-package/infer/target/*
+CLASS_PATH=$MXNET_ROOT/scala-package/assembly/$platform-$hw_type/target/*:$MXNET_ROOT/scala-package/examples/target/*:$MXNET_ROOT/scala-package/examples/target/classes/lib/*:$MXNET_ROOT/scala-package/infer/target/*
 
 # model dir and prefix
 MODEL_DIR=$1
```
Comment (lanking520) [inline scala-package/examples/scripts/infer/objectdetector/run_ssd_example.sh]: The number of MODEL_DIR should be changed after specifying the first element as CPU/GPU

File: scala-package/examples/src/main/java/org/apache/mxnetexamples/javaapi/infer/objectdetector/README.md [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+4. [Run the image inference example](#run-the-image-inference-example)
+5. [Infer APIs](#infer-api-details)
+6. [Next steps](#next-steps)
+
+
+## Prerequisites
+
+1. MXNet
+2. MXNet Scala Package
+3. [IntelliJ IDE (or alternative IDE) project setup](http://mxnet.incubator.apache.org/tutorials/scala/mxnet_scala_on_intellij.html) with the MXNet Scala Package
```
Comment (lanking520) [inline scala-package/examples/src/main/java/org/apache/mxnetexamples/javaapi/infer/objectdetector/README.md]: Update the link with new Java API

File: scala-package/examples/src/main/java/org/apache/mxnetexamples/javaapi/infer/objectdetector/README.md [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+4. wget
+
+
+## Setup Guide
+
+### Download Artifacts
+#### Step 1
+You can download the files using the script `get_ssd_data.sh`. It will download and place the model files in a `model` folder and the test image files in a `image` folder in the current directory.
+From the `scala-package/examples/scripts/infer/imageclassifier/` folder run:
```
Comment (lanking520) [inline scala-package/examples/src/main/java/org/apache/mxnetexamples/javaapi/infer/objectdetector/README.md]: imageclassifier -> objectdetector?

File: scala-package/examples/src/main/java/org/apache/mxnetexamples/javaapi/infer/objectdetector/README.md [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+
+```bash
+./get_ssd_data.sh
+```
+
+**Note**: You may need to run `chmod +x get_resnet_data.sh` before running this script.
```
Comment (lanking520) [inline scala-package/examples/src/main/java/org/apache/mxnetexamples/javaapi/infer/objectdetector/README.md]: ssd_data

File: scala-package/examples/src/main/java/org/apache/mxnetexamples/javaapi/infer/objectdetector/README.md [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+
+| Argument                      | Comments                                 |
+| ----------------------------- | ---------------------------------------- |
+| `model-path-prefix`                   | Folder path with prefix to the model (including json, params, and any synset file). |
+| `input-image`                 | The image to run inference on. |
+| `input-dir`                   | The directory of images to run inference on. |
+
+
+## How to Run Inference
+After the previous steps, you should be able to run the code using the following script that will pass all of the required parameters to the Infer API.
```
Comment (lanking520) [inline scala-package/examples/src/main/java/org/apache/mxnetexamples/javaapi/infer/objectdetector/README.md]: Java Infer API

File: scala-package/examples/src/main/java/org/apache/mxnetexamples/javaapi/infer/objectdetector/README.md [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+(Coord:,155.95807,149.96362,383.8369,418.94513)
+Class: dog
+Probabilties: 0.8226818
+(Coord:,83.82353,179.13998,206.63783,476.7875)
+```
+the outputs come from the the input image, with top3 predictions picked.
+
+
+## Infer API Details
+This example uses ObjectDetector class provided by MXNet's scala package Infer APIs. It provides methods to load the images, create NDArray out of Java BufferedImage and run prediction using Classifier and Predictor APIs.
```
Comment (lanking520) [inline scala-package/examples/src/main/java/org/apache/mxnetexamples/javaapi/infer/objectdetector/README.md]: Java Package?

File: scala-package/examples/src/main/java/org/apache/mxnetexamples/javaapi/infer/objectdetector/README.md [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+
+## References
+This documentation used the model and inference setup guide from the [MXNet Model Server SSD example](https://github.com/awslabs/mxnet-model-server/blob/master/examples/ssd/README.md).
+
+
+## Next Steps
+
+Check out the following related tutorials and examples for the Infer API:
+
+* [Image Classification with the MXNet Scala Infer API](../imageclassifier/README.md)
```
Comment (lanking520) [inline scala-package/examples/src/main/java/org/apache/mxnetexamples/javaapi/infer/objectdetector/README.md]: Java

File: scala-package/examples/scripts/infer/objectdetector/run_ssd_example.sh [hunk 8]
```diff
@@ -17,9 +17,21 @@
 # specific language governing permissions and limitations
 # under the License.
 
+hw_type=cpu
+if [[ $1 = gpu ]]
```
Comment (andrewfayres) [inline scala-package/examples/scripts/infer/objectdetector/run_ssd_example.sh]:
```suggestion
if [[ $4 = gpu ]]
```

File: Makefile [hunk 9]
```diff
@@ -605,7 +605,7 @@ scalaclean:
 
 scalapkg:
 	(cd $(ROOTDIR)/scala-package; \
-		mvn package -P$(SCALA_PKG_PROFILE),$(SCALA_VERSION_PROFILE) -Dcxx="$(CXX)" \
+		mvn package -P$(SCALA_PKG_PROFILE),$(SCALA_VERSION_PROFILE),integrationtest -Dcxx="$(CXX)" \
```
Comment (yzhliu) [inline Makefile]: why adding test here?

File: docs/tutorials/java/mxnet_java_on_intellij.md [hunk 10]
```diff
@@ -0,0 +1,210 @@
+# Run MXNet Java Examples Using the IntelliJ IDE (macOS)
+
+This tutorial guides you through setting up a simple Java project in IntelliJ IDE on macOS and demonstrates usage of the MXNet Java APIs. 
+
+## Prerequisites:
+To use this tutorial you need the following pre-requisites:
+
+- [Java 8 JDK](http://www.oracle.com/technetwork/java/javase/downloads/index.html)
```
Comment (yzhliu) [inline docs/tutorials/java/mxnet_java_on_intellij.md]: can we support java 7?

File: docs/tutorials/java/mxnet_java_on_intellij.md [hunk 11]
```diff
@@ -0,0 +1,210 @@
...
+
+Review the project's properties. The settings can be left as their default.
+
+TODO
+![project location](https://raw.githubusercontent.com/dmlc/web-data/master/mxnet/java/intellij-project-location.png)
+
+Set the project's location. The rest of the settings can be left as their default.
+
+TODO
+![project 1](https://raw.githubusercontent.com/dmlc/web-data/master/mxnet/java/intellij-project-pom.png)
```
Comment (yzhliu) [inline docs/tutorials/java/mxnet_java_on_intellij.md]: what are these TODO for?

File: docs/tutorials/scala/mxnet_java_install_and_run_examples.md [hunk 12]
```diff
@@ -0,0 +1,123 @@
...
+import org.apache.mxnet.javaapi.NDArray;
+
+public class JavaSample {
+public static void main(String[] args) {
+  System.out.println("Hello");
+  NDArray nd = NDArray.ones(Context.cpu(), new int[] {10, 20});
+
+  System.out.println("Shape of NDarray is : "  + nd.shape());
+}
+}
```
Comment (yzhliu) [inline docs/tutorials/scala/mxnet_java_install_and_run_examples.md]: indent

File: docs/tutorials/scala/mxnet_java_install_and_run_examples.md [hunk 13]
```diff
@@ -0,0 +1,123 @@
...
+
+public class App {
+public static void main(String[] args) {
+  System.out.println("Hello");
+  NDArray nd = NDArray.ones(Context.cpu(), new int[] {10, 20});
+
+  System.out.println("Shape of NDarray is : "  + nd.shape());
+
+}
+}
```
Comment (yzhliu) [inline docs/tutorials/scala/mxnet_java_install_and_run_examples.md]: indent

File: scala-package/core/src/main/scala/org/apache/mxnet/javaapi/NDArray.scala [hunk 14]
```diff
@@ -0,0 +1,387 @@
...
+  * NDArray object in mxnet.
+  * NDArray is basic ndarray/Tensor like data structure in mxnet. <br />
+  * <b>
+  * NOTE: NDArray is stored in native memory. Use NDArray in a try-with-resources() construct
+  * or a [[ResourceScope]] in a try-with-resource to have them automatically disposed. You can
+  * explicitly control the lifetime of NDArray by calling dispose manually. Failure to do this
+  * will result in leaking native memory.
+  * </b>
+  */
+class NDArray(val nd : org.apache.mxnet.NDArray ) {
```
Comment (yzhliu) [inline scala-package/core/src/main/scala/org/apache/mxnet/javaapi/NDArray.scala]: code style: remove space before `:`

File: scala-package/core/src/main/scala/org/apache/mxnet/javaapi/NDArray.scala [hunk 15]
```diff
@@ -0,0 +1,387 @@
...
+    */
+  def dispose() : Unit = nd.dispose()
+
+  /**
+    * Dispose all NDArrays who help to construct this array. <br />
+    * e.g. (a * b + c).disposeDeps() will dispose a, b, c (including their deps) and a * b
+    * @return this array
+    */
+  def disposeDeps() : NDArray = nd.disposeDepsExcept()
+  // def disposeDepsExcept(arr : Array[NDArray]) : NDArray = nd.disposeDepsExcept()
```
Comment (yzhliu) [inline scala-package/core/src/main/scala/org/apache/mxnet/javaapi/NDArray.scala]: remove or uncomment

File: scala-package/core/src/main/scala/org/apache/mxnet/javaapi/NDArray.scala [hunk 16]
```diff
@@ -0,0 +1,387 @@
...
+    * @return Current NDArray
+    */
+  def set(value : Float) : NDArray = nd.set(value)
+  def set(other : NDArray) : NDArray = nd.set(other)
+  def set(other : Array[Float]) : NDArray = nd.set(other)
+
+  def add(other : NDArray) : NDArray = this.nd + other.nd
+  def add(other : Float) : NDArray = this.nd + other
+  def _add(other : NDArray) : NDArray = this.nd += other
+  def _add(other : Float) : NDArray = this.nd += other
```
Comment (yzhliu) [inline scala-package/core/src/main/scala/org/apache/mxnet/javaapi/NDArray.scala]: underline usually means private / internal functions, and the name itself does not show the functionality. consider name it `addInplace`.

File: scala-package/examples/src/main/java/org/apache/mxnetexamples/javaapi/infer/objectdetector/SSDClassifierExample.java [hunk 17]
```diff
@@ -0,0 +1,199 @@
...
+// scalastyle:on
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+import java.io.File;
+
+public class SSDClassifierExample {
+	@Option(name = "--model-path-prefix", usage = "input model directory and prefix of the model")
```
Comment (yzhliu) [inline scala-package/examples/src/main/java/org/apache/mxnetexamples/javaapi/infer/objectdetector/SSDClassifierExample.java]: to be consistent, 4-space indent.

File: scala-package/examples/src/main/java/org/apache/mxnetexamples/javaapi/infer/objectdetector/SSDClassifierExample.java [hunk 18]
```diff
@@ -0,0 +1,199 @@
...
+		}
+		
+		try {
+			Shape inputShape = new Shape(new int[] {1, 3, 512, 512});
+			Shape outputShape = new Shape(new int[] {1, 6132, 6});
+			
+			
+			int width = inputShape.get(2);
+			int height = inputShape.get(3);
+			String outputStr = "\n";
```
Comment (yzhliu) [inline scala-package/examples/src/main/java/org/apache/mxnetexamples/javaapi/infer/objectdetector/SSDClassifierExample.java]: better to use `StringBuilder`

File: scala-package/examples/src/main/java/org/apache/mxnetexamples/javaapi/infer/objectdetector/SSDClassifierExample.java [hunk 19]
```diff
@@ -0,0 +1,199 @@
...
+		}
+		System.exit(0);
+	}
+	
+	static Boolean checkExist(List<String> arr)  {
+		Boolean exist = true;
+		for (String item : arr) {
+			exist = new File(item).exists() && exist;
+			if (!exist) {
+				logger.error("Cannot find: " + item);
```
Comment (yzhliu) [inline scala-package/examples/src/main/java/org/apache/mxnetexamples/javaapi/infer/objectdetector/SSDClassifierExample.java]: incorrect functionality - the purpose is to log each non-exist item, but with this implement, once one item does not exist, all the following items will be logged as non-exist.

File: scala-package/examples/src/main/java/org/apache/mxnetexamples/javaapi/infer/objectdetector/SSDClassifierExample.java [hunk 20]
```diff
@@ -0,0 +1,199 @@
...
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import org.apache.mxnet.javaapi.*;
+import org.apache.mxnet.infer.javaapi.ObjectDetector;
+
+// scalastyle:off
+import java.awt.image.BufferedImage;
+// scalastyle:on
```
Comment (yzhliu) [inline scala-package/examples/src/main/java/org/apache/mxnetexamples/javaapi/infer/objectdetector/SSDClassifierExample.java]: what style error needs to be turned off?

File: scala-package/mxnet-demo/java-demo/src/main/java/sample/ObjectDetection.java [hunk 21]
```diff
@@ -0,0 +1,101 @@
...
+            context.add(Context.cpu());
+        }
+        downloadModelImage();
+        Shape inputShape = new Shape(new int[] {1, 3, 512, 512});
+        Shape outputShape = new Shape(new int[] {1, 6132, 6});
+        int width = inputShape.get(2);
+        int height = inputShape.get(3);
+        List<List<ObjectDetectorOutput>> output
+                = runObjectDetectionSingle(modelPath, imagePath, context);
+        String outputStr = "\n";
```
Comment (yzhliu) [inline scala-package/mxnet-demo/java-demo/src/main/java/sample/ObjectDetection.java]: better to use StringBuilder

File: scala-package/examples/src/main/java/org/apache/mxnetexamples/javaapi/infer/objectdetector/README.md [hunk 22]
```diff
@@ -0,0 +1,116 @@
...
+- `Accuracy`
+- `Xmin`
+- `Ymin`
+- `Xmax`
+- `Ymax`
+
+
+### Setup Datapath and Parameters
+#### Step 2
+The code `Line 31: val baseDir = System.getProperty("user.dir")` in the example will automatically searches the work directory you have defined. Please put the files in your [work directory](https://stackoverflow.com/questions/16239130/java-user-dir-property-what-exactly-does-it-mean). <!-- how do you define the work directory? -->
```
Comment (nswamy) [inline scala-package/examples/src/main/java/org/apache/mxnetexamples/javaapi/infer/objectdetector/README.md]: what is this line 31 referring to? does the user have to know?

Have you tested this? When I tested this last time, it didn't work for me? 

Do you have to do this if you get the packages from [Maven](https://maven.apache.org/guides/mini/guide-3rd-party-jars-local.html).

File: scala-package/infer/src/main/scala/org/apache/mxnet/infer/javaapi/ObjectDetector.scala [hunk 23]
```diff
@@ -0,0 +1,128 @@
...
+  *                           Example: file://model-dir/ssd_resnet50_512 (containing
+  *                           ssd_resnet50_512-symbol.json, ssd_resnet50_512-0000.params,
+  *                           and synset.txt)
+  * @param inputDescriptors   Descriptors defining the input node names, shape,
+  *                           layout and type parameters
+  * @param contexts           Device contexts on which you want to run inference.
+  *                           Defaults to CPU.
+  * @param epoch              Model epoch to load; defaults to 0
+  */
+class ObjectDetector(val objDetector: org.apache.mxnet.infer.ObjectDetector){
```
Comment (nswamy) [inline scala-package/infer/src/main/scala/org/apache/mxnet/infer/javaapi/ObjectDetector.scala]: can we make this constructor private?

File: scala-package/infer/src/main/scala/org/apache/mxnet/infer/javaapi/Predictor.scala [hunk 24]
```diff
@@ -0,0 +1,86 @@
...
+  *                            layout and type parameters
+  *                            <p>Note: If the input Descriptors is missing batchSize
+  *                            ('N' in layout), a batchSize of 1 is assumed for the model.
+  * @param contexts            Device contexts on which you want to run inference; defaults to CPU
+  * @param epoch               Model epoch to load; defaults to 0
+
+  */
+
+// JavaDoc description of class to be updated in https://issues.apache.org/jira/browse/MXNET-1178
+class Predictor(val predictor: org.apache.mxnet.infer.Predictor){
```
Comment (nswamy) [inline scala-package/infer/src/main/scala/org/apache/mxnet/infer/javaapi/Predictor.scala]: constructor private? Please make all the constructors that uses the Scala Object as private

File: scala-package/examples/src/main/java/org/apache/mxnetexamples/javaapi/infer/objectdetector/SSDClassifierExample.java (hunk 20 above)
Comment (andrewfayres) [reply scala-package/examples/src/main/java/org/apache/mxnetexamples/javaapi/infer/objectdetector/SSDClassifierExample.java]: [IllegalImportChecker](https://github.com/apache/incubator-mxnet/blob/b8e36e0771ecd42137b8d5755ad638bb06026cc7/scala-package/scalastyle-config.xml#L164). It is also turned off for this reason in the non-Java version of this example [here](https://github.com/apache/incubator-mxnet/blob/b8e36e0771ecd42137b8d5755ad638bb06026cc7/scala-package/examples/src/main/scala/org/apache/mxnetexamples/infer/objectdetector/SSDClassifierExample.scala#L19)

File: scala-package/examples/src/main/java/org/apache/mxnetexamples/javaapi/infer/objectdetector/SSDClassifierExample.java (hunk 19 above)
Comment (andrewfayres) [reply scala-package/examples/src/main/java/org/apache/mxnetexamples/javaapi/infer/objectdetector/SSDClassifierExample.java]: This is also wrong in the existing scala example so fixing it there too.

**Comment by lanking520** [comment]
We are close to the end (remaining PRs)!:
https://github.com/apache/incubator-mxnet/pull/13295
https://github.com/apache/incubator-mxnet/pull/13201
https://github.com/apache/incubator-mxnet/pull/13237

File: scala-package/examples/src/main/java/org/apache/mxnetexamples/javaapi/infer/predictor/PredictorExample.java [hunk 25]
```diff
@@ -0,0 +1,200 @@
...
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * This Class is a demo to show how users can use Predictor APIs to do
+ * Image Classification with all hand-crafted Pre-processing.
+ * All helper functions for image pre-processing are
+ * currently available in ObjectDetector class.
+ */
+public class PredictorExample {
```
Comment (nswamy) [inline scala-package/examples/src/main/java/org/apache/mxnetexamples/javaapi/infer/predictor/PredictorExample.java]: this code looks like ImageClassifier. This example should show probably a simple example such as Linear Regression.

**Comment by nswamy** [comment]
Please create a follow up PR for
1) Reducing the number of instructions in the IntelliJ Setup guide
2) Add a ImageClassifier - its odd that we have a ObjectDetector but not a ImageClassifier for CV tasks.

**Pull request merged**